brave/brave-core
# Open Leo from search

Resolves https://github.com/brave/brave-browser/issues/41711

https://github.com/user-attachments/assets/bbfce0c6-3cc5-4aac-b301-e8e887fd7bdb

UI was updated on search side after the above demo video was recorded, it now looks like this, with a "Save chat in Brave Leo" at upper-right corner:
<img width="1608" alt="Screenshot 2024-10-30 at 4 13 46 PM" src="https://github.com/user-attachments/assets/e7f42d3d-87da-4cf8-bfd5-581727cad181">

- Add a new permission for opening Leo from Brave Search
- Add a navigation throttle to intercept the request for opening Leo from Brave Search (specifically, https://search.brave.com/leo#noncevalue). If the request is from Brave Search (last committed URL) and nonce is valid (identical between the one in URL and the one in nonce property), a permission prompt will be shown by default and Leo side panel would be opened with entries from Brave Search conversations staged in Leo side panel.
- A side change which removes the limitation of only have staged entries at the beginning of the conversation to accommodate that user could choose to click continue with Leo button in search UI multiple times and not only when there's empty Leo chat history. When need to remove staged entries, for example, when page context is unlinked, all staged entries will be cleared instead of whole chat history.

See Requirements `#2` and `#3` in https://docs.google.com/document/d/1idelFPpUEcKDNcyKYf3M5yw91tuIddjrlYfpaWEJjNk/edit?tab=t.0 for reference. 

S&P review: https://github.com/brave/reviews/issues/1776



## Submitter Checklist:

  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`

## Reviewer Checklist:


## After-merge Checklist:

  changes has landed on

## Test Plan:
1. Go to search.brave.com with experimental flags enabled.
2. Type a query and click `Answer with AI`
3. Type a follow-up question in search conversation mode UI
4. Once answer is generated, click "Save chat in Brave Leo" at the upper-right corner.
5. Check permission prompt is shown, and Leo side panel should be opened when permission is allowed, entries from search.brave.com should be staged in Leo.
6. Visit brave://settings/content -> Additional permissions -> Leo AI chat, search.brave.com should be in the allow list.
7. Remove the allow entry in settings UI. 
8. Open a new tab and repeat step 1 to 4.
9. Permission prompt should be shown again, reject the prompt, Leo panel should not be opened.

Test opt-out case:
Try the feature in a profile that hasn't opted-in to Leo, when the continue with Leo button is clicked, accept the permission prompt, Leo should be opened with opt-in screen with disclaimer, search entries should appear after opt-in.

Test private window:
Try the feature in private window, shouldn't have any actions such as opening Leo or permission prompt.
